File: tests/fixtures/java/AuthService.java
package com.example.auth;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

// Target: bookmark an interface
interface AuthProvider {
    Claims validateToken(String token) throws AuthException;
    String refreshToken(String token) throws AuthException;
}

// Target: bookmark a record/class
class Claims {
    final String subject;
    final long expiry;
    final List<String> roles;

    Claims(String subject, long expiry, List<String> roles) {
        this.subject = subject;
        this.expiry = expiry;
        this.roles = roles;
    }
}

// Target: bookmark an enum
enum AuthError {
    INVALID_TOKEN,
    EXPIRED,
    INSUFFICIENT_PERMISSIONS
}

// Target: bookmark an exception class
class AuthException extends Exception {
    private final AuthError error;

    AuthException(AuthError error, String message) {
        super(message);
        this.error = error;
    }

    AuthError getError() {
        return error;
    }
}

// Target: bookmark a class
public class AuthService implements AuthProvider {
    private final String secret;
    private final String issuer;
    private final Map<String, Claims> tokenCache = new HashMap<>();

    public AuthService(String secret, String issuer) {
        this.secret = secret;
        this.issuer = issuer;
    }

    // Target: bookmark a method
    @Override
    public Claims validateToken(String token) throws AuthException {
        Claims cached = tokenCache.get(token);
        if (cached != null) {
            if (cached.expiry <= System.currentTimeMillis()) {
                tokenCache.remove(token);
                throw new AuthException(AuthError.EXPIRED, "Token expired");
            }
            return cached;
        }

        Claims claims = decode(token);
        tokenCache.put(token, claims);
        return claims;
    }

    @Override
    public String refreshToken(String token) throws AuthException {
        Claims claims = decode(token);
        return encode(claims);
    }

    // Target: bookmark a private method
    private Claims decode(String token) throws AuthException {
        if (token == null || token.isEmpty()) {
            throw new AuthException(AuthError.INVALID_TOKEN, "Invalid token");
        }
        return new Claims("user-1", System.currentTimeMillis() + 3600000, List.of("admin"));
    }

    private String encode(Claims claims) {
        return issuer + ":" + claims.subject + ":" + secret;
    }

    // Target: bookmark a method with complex params
    public void checkPermission(Claims claims, String required, String resource)
            throws AuthException {
        if (!claims.roles.contains(required)) {
            throw new AuthException(
                AuthError.INSUFFICIENT_PERMISSIONS,
                required + " on " + resource
            );
        }
    }

    // Target: bookmark a static factory
    public static AuthService createDefault() {
        return new AuthService("default-secret", "codemark");
    }
}

File: tests/fixtures/kotlin/AuthService.kt
package com.example.auth

// Target: bookmark a data class
data class Claims(
    val subject: String,
    val expiry: Long,
    val roles: List<String>
)

// Target: bookmark a sealed class
sealed class AuthError(val message: String) {
    object InvalidToken : AuthError("invalid token")
    object Expired : AuthError("token expired")
    data class InsufficientPermissions(val required: String) :
        AuthError("insufficient permissions: $required")
}

// Target: bookmark an interface
interface AuthProvider {
    fun validateToken(token: String): Claims
    fun refreshToken(token: String): String
}

// Target: bookmark a class
class AuthService(
    private val secret: String,
    private val issuer: String = "codemark"
) : AuthProvider {

    private val tokenCache = mutableMapOf<String, Claims>()

    // Target: bookmark a method
    override fun validateToken(token: String): Claims {
        tokenCache[token]?.let { cached ->
            if (cached.expiry <= System.currentTimeMillis()) {
                tokenCache.remove(token)
                throw IllegalStateException(AuthError.Expired.message)
            }
            return cached
        }

        val claims = decode(token)
        tokenCache[token] = claims
        return claims
    }

    override fun refreshToken(token: String): String {
        val claims = decode(token)
        return encode(claims)
    }

    // Target: bookmark a private method
    private fun decode(token: String): Claims {
        if (token.isEmpty()) {
            throw IllegalArgumentException(AuthError.InvalidToken.message)
        }
        return Claims(
            subject = "user-1",
            expiry = System.currentTimeMillis() + 3600000,
            roles = listOf("admin")
        )
    }

    private fun encode(claims: Claims): String {
        return "$issuer:${claims.subject}:$secret"
    }

    // Target: bookmark a method with complex params
    fun checkPermission(claims: Claims, required: String, resource: String) {
        if (required !in claims.roles) {
            throw SecurityException(
                AuthError.InsufficientPermissions(required).message
            )
        }
    }

    companion object {
        // Target: bookmark a companion object factory
        fun createDefault(): AuthService {
            return AuthService("default-secret")
        }
    }
}

// Target: bookmark a top-level function
fun createDefaultAuthService(): AuthService {
    return AuthService.createDefault()
}
